File: gtci/15-two-heaps/python3/3-maximise-capital.py
import heapq


class Solution:
    def findMaximumCapital(self, capital, profits, numberOfProjects, initialCapital):
        self.max_heap = (
            []
        )  # max heap on profits -- all the projects we can afford are in the heap

        projects = sorted(zip(capital, profits), key=lambda x: x[0])
        capital, profits = zip(*projects)

        n_projects_chosen = 0
        capital_available = initialCapital
        next_project_to_consider = 0
        while n_projects_chosen < numberOfProjects:
            # put all the projects we can afford into the heap
            next_project_to_consider = self.populate_heap(
                next_project_to_consider, capital, profits, capital_available
            )

            # choose the most profitable project we can afford
            if not self.max_heap:
                break

            profit, project = heapq.heappop(self.max_heap)
            profit = -profit

            capital_available += profit
            n_projects_chosen += 1

        return capital_available

    def populate_heap(
        self, next_project_to_consider, capital, profits, capital_available
    ):
        project = next_project_to_consider
        while project < len(capital) and capital[project] <= capital_available:
            heapq.heappush(self.max_heap, (-profits[project], project))
            project += 1

        return project


if __name__ == "__main__":
    solution = Solution()

    capital = [0, 1, 2]
    profits = [1, 2, 3]
    numberOfProjects = 2
    initialCapital = 1
    assert (
        solution.findMaximumCapital(capital, profits, numberOfProjects, initialCapital)
        == 6
    )

    capital = [0, 1, 2, 3]
    profits = [1, 2, 3, 5]
    numberOfProjects = 3
    initialCapital = 0
    assert (
        solution.findMaximumCapital(capital, profits, numberOfProjects, initialCapital)
        == 8
    )

    print("All test cases passed.")

File: gtci/15-two-heaps/cpp/3-maximise-capital.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Solution {
   public:
    std::priority_queue<int> maxHeap;
    int findMaximumCapital(
      const vector<int>& capital,
      const vector<int>& profits,
      int numberOfProjects,
      int initialCapital
    ) {
      std::vector<std::pair<int, int>> projects;
      for (size_t i = 0; i < capital.size(); i++)
        projects.push_back({capital[i], profits[i]});
      
      std::sort(projects.begin(), projects.end(), [](auto& a, auto& b) { return a.first < b.first; });

      int capitalAvailable = initialCapital;
      int nextProjectToConsider = 0;
      
      for (int i = 0; i < numberOfProjects; i++) {
        nextProjectToConsider = populateMaxHeap(nextProjectToConsider, projects, capitalAvailable);

        if (maxHeap.empty())
          break;
        
        int profit = maxHeap.top();
        maxHeap.pop();

        capitalAvailable += profit;
      }

      return capitalAvailable;
    }

    int populateMaxHeap(
      const int& nextProjectToConsider,
      const std::vector<std::pair<int, int>>& projects,
      const int& availableCapital
    ) {
      int project = nextProjectToConsider;
      while (project < projects.size() && projects[project].first <= availableCapital) {
        maxHeap.push(projects[project].second);
        project++;
      }

      return project;
    }
};

int main() {
    Solution solution;

    assert(solution.findMaximumCapital({0, 1, 2}, {1, 2, 3}, 2, 1) == 6);
    assert(solution.findMaximumCapital({0, 1, 2, 3}, {1, 2, 3, 5}, 3, 0) == 8);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}
